### PR TITLE
Songs index

### PR DESCRIPTION
**Pull request by dgrosenblatt**
This pull request adds an endpoint for GET /songs that responds with JSON for all songs.

Sample request:

```
GET /songs HTTP/1.1
Accept: application/vnd.spotify-apprentice-app.com; version=1
```

Sample response:

```
HTTP/1.1 200 OK
{
  "songs": [
    {
      "id": "7e4ea5d9-3e29-446f-bd0d-3c8c5064f524",
      "title": "No Surprises",
      "artist": "Radiohead",
      "album": "OK Computer",
      "duration": 192
    }
  ]
}
```

### CODE REVIEW

File: db/migrate/20150609191815_create_songs.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CreateSongs < ActiveRecord::Migration
+  def change
+    enable_extension 'uuid-ossp'
```
Comment (dthyresson) [inline db/migrate/20150609191815_create_songs.rb]: Are you certain you need to enable this again?

The `schema.db` in `master` should already account for this:

```
ActiveRecord::Schema.define(version: 20150609174024) do

  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"
  enable_extension "uuid-ossp"
```

... so, no need to enable it in each migration.

File: spec/models/song_spec.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+require 'rails_helper'
+
+describe Song do
+  it { should validate_presence_of :title }
+  it { should validate_presence_of :artist }
+  it { should validate_presence_of :album }
+  it { should validate_presence_of :duration }
```
Comment (dthyresson) [inline spec/models/song_spec.rb]: All these validations are spot on and really doesn't need much more.

But ... can you think of any other validation(s) you might add, say on `:duration` that could be useful?

File: spec/requests/v1/songs_request_spec.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+describe 'songs endpoints' do
+  describe 'GET /index' do
+    it 'returns JSON for all songs' do
```
Comment (dthyresson) [inline spec/requests/v1/songs_request_spec.rb]: FYI - often, when having multiple test cases, it can be helpful to wrap them in a 

```
context 'when there are several songs` do
  it 'returns JSON for all songs' do
    ...
  end
end
```

and

```
context 'when there are no songs' do
...
end
```

To give ... context ... for the scenario that the expectation is validating.

**Comment by dthyresson** [comment]
Looks great. Nice work. Will let helen review as well, but looks good to :shipit:

File: spec/requests/v1/artists_request_spec.rb [hunk 4]
```diff
@@ -6,7 +6,6 @@
       artists = create_list(:artist, 3)
 
       get(artists_url, {}, accept_headers)
-binding.pry
```
Comment (hchood) [inline spec/requests/v1/artists_request_spec.rb:9]: Derp.  :+1:

File: app/models/song.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class Song < ActiveRecord::Base
+  validates :title,
+    presence: true
```
Comment (hchood) [inline app/models/song.rb]: Generally put these on one line until it gets (close) to 80 characters long.

File: db/schema.rb [hunk 6]
```diff
@@ -11,13 +11,13 @@
...
 
-ActiveRecord::Schema.define(version: 20150609174024) do
+ActiveRecord::Schema.define(version: 20150609191815) do
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
   enable_extension "uuid-ossp"
 
-  create_table "artists", id: :uuid, default: "uuid_generate_v4()", force: :cascade do |t|
+  create_table "artists", id: :uuid, default: nil, force: :cascade do |t|
```
Comment (hchood) [inline db/schema.rb]: Yeah I think enabling the UUID extension again screws up the UUIDs on artist - the default here should be `uuid_generate_v4()` 

http://www.postgresql.org/docs/9.4/static/uuid-ossp.html:

> uuid_generate_v4()    | This function generates a version 4 UUID, which is derived entirely from random numbers.

File: db/migrate/20150609191815_create_songs.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+    create_table :songs, id: :uuid do |t|
+      t.timestamps null: false
+
+      t.string :title
+      t.string :artist
+      t.string :album
+      t.integer :duration
+    end
+  end
```
Comment (hchood) [inline db/migrate/20150609191815_create_songs.rb:12]: Could think about a uniqueness constraint on some combination of these attributes, just so you don't get duplicate records.  Maybe `title, album`?

File: spec/factories/songs.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :song do
+    title 'Paranoid Android'
+    artist 'Radiohead'
+    album 'OK Computer'
+    duration 325
+  end
+
```
Comment (hchood) [inline spec/factories/songs.rb]: Solid factory choice. :+1: 

Can get rid of this extra blank line.

File: spec/models/song_spec.rb (hunk 2 above)
Comment (hchood) [inline spec/models/song_spec.rb]: Also I usually nest my validations inside a `describe 'validations'` block for organizational purposes.  (Similarly with associations in a `describe 'associations'`

File: db/migrate/20150609191815_create_songs.rb (hunk 7 above)
Comment (hchood) [inline db/migrate/20150609191815_create_songs.rb:12]: Also probably want a `spotify_uri` :)

File: spec/models/song_spec.rb (hunk 2 above)
Comment (hchood) [inline spec/models/song_spec.rb]: Also if you added a uniqueness constraint to the DB, would want a matching validation here.

File: spec/requests/v1/songs_request_spec.rb (hunk 3 above)
Comment (hchood) [inline spec/requests/v1/songs_request_spec.rb]: Yes, exactly.

Particularly when you get to multiple edge case tests, this is very useful.  For example, if tests out an endpoint to add a song, might have context blocks like:

``` ruby
context 'authenticated user' do
  context 'with valid params' do
    # ...
  end

  context 'missing required attributes' do
    # ...
  end

  context 'when song already exists' do
    # ...
  end
end

context 'unauthenticated user' do
  # ...
end
```

File: spec/requests/v1/songs_request_spec.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      songs = create_list(:song, 10)
+      get(songs_url, {}, accept_headers)
+      expect(response).to have_http_status :ok
+      expect(response).to match_response_schema :songs
+    end
```
Comment (hchood) [inline spec/requests/v1/songs_request_spec.rb:13]: I usually put some blank lines in here between the setup & the expectations, just for readability.

**Comment by hchood** [comment]
Agree with DT. :shipit: !

:beers: on first PR.

File: spec/models/song_spec.rb (hunk 2 above)
Comment (dgrosenblatt) [inline spec/models/song_spec.rb]: Added numericality and > 0 for duration. Added uniqueness for title scoped to album. @hchood @dthyresson

File: app/models/song.rb (hunk 5 above)
Comment (dgrosenblatt) [inline app/models/song.rb]: Fine :angry:

File: db/schema.rb (hunk 6 above)
Comment (dgrosenblatt) [inline db/schema.rb]: You were right-- it did. Removed enable_extension from the migration.

File: spec/requests/v1/songs_request_spec.rb (hunk 9 above)
Comment (dgrosenblatt) [inline spec/requests/v1/songs_request_spec.rb:13]: Reorganized with blank lines and context blocks.

**Comment by hchood** [comment]
:shipit: :rocket:
